pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go

```diff
@@ -0,0 +1,170 @@
+			returnErr:   fmt.Errorf("conflicting plugin name"),
+		},
+	}
+
+	callbackCount := 0
+	w.AddHandler("example-plugin-type", func(name string, versions []string, sockPath string) error {
+		glog.Infof("receives plugin watcher callback for test: %s\n", testCases[callbackCount].description)
+		require.Equal(t, "example-plugin", name, "Plugin name mismatched!!")
+		require.Equal(t, []string{"v1beta1", "v1beta2"}, versions, "Plugin version mismatched!!")
+		// Verifies the grpcServer is ready to serve services.
+		conn, err := grpc.Dial(socketPath, grpc.WithInsecure(), grpc.WithBlock(),
+			grpc.WithTimeout(10*time.Second),
+			grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+				return net.DialTimeout("unix", addr, timeout)
+			}))
+		require.Nil(t, err)
+		defer conn.Close()
+
+		// The plugin handler should be able to use any listed service API version.
+		v1beta1Client := v1beta1.NewExampleClient(conn)
+		v1beta2Client := v1beta2.NewExampleClient(conn)
+
+		// Tests v1beta1 GetExampleInfo
+		_, err = v1beta1Client.GetExampleInfo(context.Background(), &v1beta1.ExampleRequest{})
+		require.Nil(t, err)
+
+		// Tests v1beta1 GetExampleInfo
+		_, err = v1beta2Client.GetExampleInfo(context.Background(), &v1beta2.ExampleRequest{})
+		require.Nil(t, err)
+		ret := testCases[callbackCount].returnErr
+		callbackCount++
+		return ret
+	})
+	err = w.Start()
+	require.NoError(t, err)
+
+	p := NewTestExamplePlugin() //first cbk
+	go func() {
+		p.Serve(socketPath)
+	}()
+	require.True(t, waitForPluginRegistrationStatus(t, p.registrationStatus))
+
+	// Trying to start a plugin service at the same socket path should fail
+	// with "bind: address already in use"
+	err = p.Serve(socketPath)
+	require.NotNil(t, err)
+
+	// grpcServer.Stop() will remove the socket and starting plugin service
+	// at the same path again should succeeds and trigger another callback.
+	p.Stop()
+	p = NewTestExamplePlugin() // 2nd cbk
+	go func() {
+		p.Serve(socketPath)
+	}()
+	require.True(t, waitForPluginRegistrationStatus(t, p.registrationStatus))
+
+	// Starting another plugin with the same name got verification error.
+	p2 := NewTestExamplePlugin() //3rd cbk
+	socketPath2 := socketDir + "/plugin2.sock"
+	go func() {
+		p2.Serve(socketPath2)
+	}()
+	require.False(t, waitForPluginRegistrationStatus(t, p2.registrationStatus))
+
+	// Restarts plugin watcher should traverse the socket directory and issues a
+	// callback for every existing socket.
+	glog.Infof("stopping watcher 1\n")
+	err = w.Stop()
+	require.NoError(t, err)
+	errCh := make(chan error)
+	go func() {
+		errCh <- w.Start()
+	}()
+	require.True(t, waitForPluginRegistrationStatus(t, p.registrationStatus))
+	require.False(t, waitForPluginRegistrationStatus(t, p2.registrationStatus))
+
+	err = <-errCh
+	require.NoError(t, err)
+
+	// This is to wait here till w.watcher has got updated in the goroutine.
+	// If we dont do this, w.Stop() will race with w.Start()
+	for {
```

Comment:
Can we have a better testing model than an infinite loop?
Having a race between start and stop usually is a good sign that the architecture (setup/teardown) needs a bit more thinking.

In this case it seems like a better model to have the watcher in the `NewWatcher` function than in the `Start` function.
Additionally the go race detector is going to detect this code as a data race.